x-pack/plugins/security_solution/server/endpoint/routes/metadata/handlers.ts

```diff
@@ -0,0 +1,290 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import Boom from 'boom';
+import { RequestHandlerContext, Logger, RequestHandler } from 'kibana/server';
+import { TypeOf } from '@kbn/config-schema';
+import {
+  HostInfo,
+  HostMetadata,
+  HostResultList,
+  HostStatus,
+  MetadataQueryStrategyVersions,
+} from '../../../../common/endpoint/types';
+import { getESQueryHostMetadataByID, kibanaRequestToMetadataListESQuery } from './query_builders';
+import { Agent, AgentStatus } from '../../../../../ingest_manager/common/types/models';
+import { EndpointAppContext, HostListQueryResult } from '../../types';
+import { GetMetadataListRequestSchema, GetMetadataRequestSchema } from './index';
+import { findAllUnenrolledAgentIds } from './support/unenroll';
+import { findAgentIDsByStatus } from './support/agent_status';
+import { EndpointAppContextService } from '../../endpoint_app_context_services';
+
+export interface MetadataRequestContext {
+  endpointAppContextService: EndpointAppContextService;
+  logger: Logger;
+  requestHandlerContext: RequestHandlerContext;
+}
+
+const HOST_STATUS_MAPPING = new Map<AgentStatus, HostStatus>([
+  ['online', HostStatus.ONLINE],
+  ['offline', HostStatus.OFFLINE],
+  ['unenrolling', HostStatus.UNENROLLING],
+]);
+
+/**
+ * 00000000-0000-0000-0000-000000000000 is initial Elastic Agent id sent by Endpoint before policy is configured
+ * 11111111-1111-1111-1111-111111111111 is Elastic Agent id sent by Endpoint when policy does not contain an id
+ */
+
+const IGNORED_ELASTIC_AGENT_IDS = [
+  '00000000-0000-0000-0000-000000000000',
```

Comment:
This looks a bit hacky to be honest. I wonder if we could find a better way to do this.

Reply:
this is done because the Endpoint will send up documents before getting the first Integration Policy from the Agent.  As a result, we'll have IDs that we can't compare against any Agents to determine if an Agent or Endpoint is unenrolled.

An alternative solution, I think, would require some form of messaging from the Endpoint itself as we will not be able to figure our which Agent the Endpoint comes from.  @ferullo had said in the past the we should continue to send this initial document, which I agree with since it can signify an important step in the lifecycle and help with debugging later.  It's just very difficult to tell when an Endpoint is unenrolled or stopped, etc.

Another long term solution would be to form our views to look for Agents first.  Then enrich the list of Agents with Endpoint data.  If we cannot find Endpoint data for that Agent, we show some message (Endpoint pending, etc...).  I think this could be a vision for more Integration specific views in Ingest.  Per Integration, we could see a list of Agents with some Integration specific (i.e. Endpoint) data enrichments.

Reply:
Even if endpoint did not receive a policy yet, I assume it already knows about the Agent id under which it is running? Or is this only sent down with the policy?

Reply:
> Even if endpoint did not receive a policy yet, I assume it already knows about the Agent id under which it is running? Or is this only sent down with the policy?

This is only sent down with the Policy.  This "initial" policy is hardcoded and ships with the Endpoint.  We don't get updated information until the Endpoint gets the first Policy from Agent.  These initial documents are sent after the ES connection is established.  Is this correct @ferullo ?  Just want to make sure I'm not misrepresenting anything

Reply:
@blakerouse @michalpristas It would be nice if we could ship down the agent id already as part of the initial "negotiation" between Agent and Endpoint.

Reply:
> It would be nice if we could ship down the agent id already as part of the initial "negotiation" between Agent and Endpoint.

Is this a step before the Agent sends the first policy to Endpoint?  Also, I could be wrong, but unless the agent ID is sent down as a part of the Policy, the Endpoint may need some extra work to echo this ID back to the Security app.

This could be a good solution, but I need to understand what the Endpoint does more deeply.

Reply:
This was a break fix solution at the last hour. If there is a longer term more viable solution then let us implement it. it will require testing and code change end to end

Reply:
@ruflin The initial configuration that we pass to Endpoint already contains the `fleet.agent.id`. At no point do we send Endpoint a configuration without `fleet.agent.id`.

Reply:
@blakerouse If I understood the above correct, endpoint already starts to send events before they receive the first policy. If that is the case, is there a way we could have the agent.id as part of the "initiation" phase?

Reply:
@ruflin I question how that is possible, being that Endpoint would not even know the elasticsearch output information until Agent sends them the configuration that already includes `fleet.agent.id`.

Reply:
> I question how that is possible, being that Endpoint would not even know the elasticsearch output information until Agent sends them the configuration that already includes `fleet.agent.id`.

@ferullo my understanding is that the Endpoint has the initial documents ready to send and this is essentially a flush?

Reply:
Yes. Endpoint caches data internally to send later when it does not have an active connection to ES.